### DISCUSSION TITLE
Choice of ORM

### DISCUSSION THREAD

**Discussion by anderslatif** [MinitwitGroupI/MiniTwit#57] [Polls]
Choice of ORM

**Comment by anderslatif** [comment]
Here is my conclusion regarding Prisma. Prisma is out of the question. It requires .prisma files and Python Prisma is not so well documented. 

The question is between:
1. Doing it with Sqlalchemy because spending time learning it will pay off in the future since it is widely used. 
https://github.com/MinitwitGroupI/ORM_Benchmarks/blob/main/sqlalchemy/models.py
2. Doing it with Peewee. Now hear me out... The way Peewee defines relations is so much nicer and they have a working "Minitwit" example that basically follows our schema even down to the usage of gravatar:
https://docs.peewee-orm.com/en/latest/peewee/example.html#example-app

**Reply by Metallist1** [reply]
Peewee looks very nice and even has good examples. I think it would be nicer to work with but I believe its more useful to have SQLAcademy as a general example for the future projects.

**Reply by mirbitu** [reply]
From my perspective, I think I am most interested in whatever is most widely used. Being unfamiliar with ORMs in general, I would prefer our use of an ORM to be what's most commonly used to have a good reference/base for future explorations with other ORMs.
From what I've researched, SQLAlchemy seems to be recognized as a good choice related to my preferences, so that would be my vote. Also helps with it having nicely available documentation and so on.
But it's worth mentioning that I have no issue with using Peewee instead. It also seems to be commonly used, as well simple to use. If we're busy with many things, it would be a perfectly fine choice.

**Reply by anderslatif** [reply]
I also want to mention that if we choose Peewee it would make sense to copy from their working example for the same schema as ours but then we must make it explicitly clear in multiple places (code base, final report).
We don't have to concern ourselves with crediting for SQLAlchemy since I wrote the models myself. 

Do note, however, that choosing SQLAlchemy will require more work and a lot of tinkering to get the queries just right.  
Here is an example of how to rewrite the timeline query:
https://github.com/MinitwitGroupI/ORM_Benchmarks/blob/main/sqlalchemy/main.py#L16

I don't have any preference either way but I'm glad that we are making a group decision since it affects everyone.
Hopefully, everyone will be involved in slowly refactoring and writing these types of queries.

**Reply by var-i** [reply]
I was all for SQL alchemy until I looked at the models.py you sent. Peewee has a much nicer model syntax. I worked a lot with Django's ORM and I think it uses Peewee under the hood. The syntax is almost identical.

However, I can't imagine that the query syntax is much different in Peewee for more complex queries such as the timeline one you linked. I think it will be pretty similar.

**Reply by anderslatif** [reply]
Peewee seems to be what I would categorize as query builders that use method chaining. 
Libraries like Linq, Jinq and Knex come to mind. 
These are libraries known for  being nice.

I completely agree with you, however. 
When your query gets complex all ORMs rear their ugly head. 
I just wish they were a little more upfront about it.
